File: service-rest/src/main/java/com/bnorm/barkeep/rest/RestRecipeService.java
// Copyright 2016 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.rest;

import java.util.Collection;
import java.util.Set;

import javax.annotation.Nullable;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RestController;

import com.bnorm.barkeep.db.DbRecipeService;
import com.bnorm.barkeep.db.DbUserService;
import com.bnorm.barkeep.model.Component;
import com.bnorm.barkeep.model.Recipe;
import com.bnorm.barkeep.model.User;
import com.bnorm.barkeep.service.RecipeService;
import com.bnorm.barkeep.service.UserService;
import com.fasterxml.jackson.annotation.JsonView;

@RestController
@RequestMapping("/recipes")
public class RestRecipeService extends AbstractRestService implements RecipeService {

  private final UserService userService;
  private final RecipeService recipeService;

  @Autowired
  public RestRecipeService(DbUserService userService, DbRecipeService recipeService) {
    this.userService = userService;
    this.recipeService = recipeService;
  }

  private Recipe findByOwner(long userId, long recipeId) {
    Recipe recipe = recipeService.getRecipe(recipeId);
    if (!isOwnedBy(recipe, userId)) {
      throw new NotFound("Unable to find recipe with id=%d", recipeId);
    }
    return recipe;
  }

  private Recipe findByUser(User user, long recipeId) {
    Recipe recipe = recipeService.getRecipe(recipeId);
    if (!user.getRecipes().contains(recipe)) {
      throw new NotFound("Unable to find recipe with id=%d", recipeId);
    }
    return recipe;
  }

  @JsonView(Recipe.class)
  @RequestMapping(method = RequestMethod.GET)
  @Override
  public Collection<Recipe> getRecipes() {
    User user = userService.getUser(currentUser().getId());
    return user.getRecipes();
  }

  @JsonView(Recipe.class)
  @RequestMapping(value = "/{recipeId}", method = RequestMethod.GET)
  @Override
  public Recipe getRecipe(@PathVariable("recipeId") long recipeId) {
    User user = userService.getUser(currentUser().getId());
    return findByUser(user, recipeId);
  }

  @JsonView(Recipe.class)
  @RequestMapping(method = RequestMethod.POST)
  @Override
  public Recipe createRecipe(Recipe recipe) {
    User currentUser = currentUser();
    if (!isOwnedBy(recipe, currentUser.getId())) {
      throw new BadRequest("Cannot create recipe owned by another user");
    }

    if (recipe.getOwner() == null) {
      return recipeService.createRecipe(new RecipeWithOwner(recipe, currentUser));
    } else {
      return recipeService.createRecipe(recipe);
    }
  }

  @JsonView(Recipe.class)
  @RequestMapping(value = "/{recipeId}", method = RequestMethod.PUT)
  @Override
  public Recipe setRecipe(@PathVariable("recipeId") long recipeId, @RequestBody Recipe recipe) {
    findByOwner(currentUser().getId(), recipeId);
    return recipeService.setRecipe(recipeId, recipe);
  }

  @RequestMapping(value = "/{recipeId}", method = RequestMethod.DELETE)
  @Override
  public void deleteRecipe(@PathVariable("recipeId") long recipeId) {
    findByOwner(currentUser().getId(), recipeId);
    recipeService.deleteRecipe(recipeId);
  }

  private static class RecipeWithOwner implements Recipe {

    private final Recipe recipe;
    private final User owner;

    private RecipeWithOwner(Recipe recipe, User owner) {
      this.recipe = recipe;
      this.owner = owner;
    }

    @Override
    public Long getId() {
      return recipe.getId();
    }

    @Nullable
    @Override
    public String getTitle() {
      return recipe.getTitle();
    }

    @Nullable
    @Override
    public String getDescription() {
      return recipe.getDescription();
    }

    @Override
    public User getOwner() {
      return owner;
    }

    @Nullable
    @Override
    public String getImageUrl() {
      return recipe.getImageUrl();
    }

    @Nullable
    @Override
    public String getInstructions() {
      return recipe.getInstructions();
    }

    @Nullable
    @Override
    public String getSource() {
      return recipe.getSource();
    }

    @Override
    public Set<Component> getComponents() {
      return recipe.getComponents();
    }
  }
}

File: service-rest/src/main/kotlin/com/bnorm/barkeep/rest/RestRecipeService.kt
// Copyright 2017 (C) BNORM Software. All rights reserved.
package com.bnorm.barkeep.rest

import com.bnorm.barkeep.db.DbRecipeService
import com.bnorm.barkeep.db.DbUserService
import com.bnorm.barkeep.model.Component
import com.bnorm.barkeep.model.Recipe
import com.bnorm.barkeep.model.User
import com.bnorm.barkeep.service.RecipeService
import com.fasterxml.jackson.annotation.JsonView
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/recipes")
class RestRecipeService
@Autowired
constructor(private val userService: DbUserService,
            private val recipeService: DbRecipeService) : AbstractRestService(), RecipeService {

  private fun findByOwner(userId: Long, recipeId: Long): Recipe {
    val recipe = recipeService.getRecipe(recipeId)
    if (!isOwnedBy(recipe, userId)) {
      throw NotFound("Unable to find recipe with id=$recipeId")
    }
    return recipe!!
  }

  private fun findByUser(user: User?, recipeId: Long): Recipe {
    val recipe = recipeService.getRecipe(recipeId)
    val recipes = user?.recipes
    if (recipe == null || recipes != null && !recipes.contains(recipe)) {
      throw NotFound("Unable to find recipe with id=$recipeId")
    }
    return recipe
  }

  @JsonView(Recipe::class)
  @RequestMapping(method = arrayOf(RequestMethod.GET))
  override fun getRecipes(): Collection<Recipe> {
    val user = userService.getUser(currentUser().id!!)
    return user!!.recipes
  }

  @JsonView(Recipe::class)
  @RequestMapping(value = "/{recipeId}", method = arrayOf(RequestMethod.GET))
  override fun getRecipe(@PathVariable("recipeId") id: Long): Recipe? {
    val user = userService.getUser(currentUser().id!!)
    return findByUser(user, id)
  }

  @JsonView(Recipe::class)
  @RequestMapping(method = arrayOf(RequestMethod.POST))
  override fun createRecipe(@RequestBody recipe: Recipe): Recipe {
    if (recipe.id != null) {
      throw BadRequest("Cannot create recipe with existing id=$recipe.id")
    }
    val currentUser = currentUser()
    if (recipe.owner == null) {
      return recipeService.createRecipe(RecipeWithOwner(recipe, currentUser))
    } else if (!isOwnedBy(recipe, currentUser.id!!)) {
      throw BadRequest("Cannot create recipe owned by another user")
    } else {
      return recipeService.createRecipe(recipe)
    }
  }

  @JsonView(Recipe::class)
  @RequestMapping(value = "/{recipeId}", method = arrayOf(RequestMethod.PUT))
  override fun setRecipe(@PathVariable("recipeId") id: Long, @RequestBody recipe: Recipe): Recipe {
    findByOwner(currentUser().id!!, id)
    return recipeService.setRecipe(id, recipe)
  }

  @RequestMapping(value = "/{recipeId}", method = arrayOf(RequestMethod.DELETE))
  override fun deleteRecipe(@PathVariable("recipeId") id: Long) {
    findByOwner(currentUser().id!!, id)
    recipeService.deleteRecipe(id)
  }

  private class RecipeWithOwner(private val recipe: Recipe, override val owner: User) : Recipe {
    override val id: Long?
      get() = recipe.id
    override val title: String?
      get() = recipe.title
    override val description: String?
      get() = recipe.description
    override val imageUrl: String?
      get() = recipe.imageUrl
    override val instructions: String?
      get() = recipe.instructions
    override val source: String?
      get() = recipe.source
    override val components: Set<Component>?
      get() = recipe.components
  }
}
